<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment</b>
    
Hi, Ken! I am a reviewer on this project. 

Before we start, I want to pay your attention to the color marking:
    
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Great solutions and ideas that can and should be used in the future are in green comments.   
</div>    
    
    
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>

Yellow color indicates what should be optimized. This is not necessary, but it will be great if you make changes to this project.
</div>      
    
    
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>

Issues that need to be corrected to get right results are indicated in red comments. Note that the project cannot be accepted until these issues are resolved.
</div>    

<hr>
    
**Please, use some color other than those listed to highlight answers to my comments.**
I would also ask you **not to change, move or delete my comments** so that it would be easier for me to navigate during the next review.
    
In addition, my comments are defined as headings. 
They can mess up the content, however, they are convenient, since you can immediately go to them. I will remove the headings from my comments in the next review. 
   
    
    
<hr>
    
<font color='dodgerblue'>**A few words about the project:**</font> you did a great job, everything is clear and neat. However, there are some issues that need to be fixed. I've also  left some comments with the recommendations for improving the project. If you have any questions, feel free to ask me.
    
    
I will wait for the project for a second review :)
    
    
📌 Here are some tips and tricks that may help you with the Markdown cells:    
<hr style="border-top: 3px solid purple; "></hr>

You can leave comments using this code inside a Markdown cell:
    
    
    <div class="alert alert-info">
    <h2> Student's comment</h2>

    Your text here. 
    </div>

    
    
    <font color='red'> This code is used to change text color. </font>     

<font color='red'> It will look like this. </font> 
    
If you don't want your comments to be headings, replace **h2** with **b** or just add `<a class="tocSkip">` after the phrase *Student's comment*.


You can find out how to **format text** in a Markdown cell or how to **add links** [here](https://sqlbak.com/blog/jupyter-notebook-markdown-cheatsheet) и [and here](https://medium.com/analytics-vidhya/the-ultimate-markdown-guide-for-jupyter-notebook-d5e5abf728fd).



</div>

<div class="alert alert-info">
    <h2>Question about accessibility</h2>

    Hey,
    
    I noticed the "alert-success" and "alert-failure" have a color scheme that's a possible accessiblity issue. As a colorblind user, those particular shades of green and red are difficult to distinguish from each other. 
    
    Luckily in this case I can just look for those respective classes in the markdown, but it's still a little frustrating.
</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 2</b>
    

Hi! Thank you for letting me know that. I updated my comments and added specific symbols: 
    
    
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ✔️</b>
    
Great solutions and ideas that can and should be used in the future are in green comments.   
</div>    
    
    
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ⚠️</b>

Yellow color indicates what should be optimized. This is not necessary, but it will be great if you make changes to this project.
</div>      
    
    
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ❌</b>

Issues that need to be corrected to get right results are indicated in red comments. Note that the project cannot be accepted until these issues are resolved.
</div>    




I hope it helps. I've left new comments titled as **Reviewer's comment 2**. 
    
    
Your project has passed code review. Congratulations 😊
 
    
This is a good [article](https://www.kaggle.com/ramprakasism/pandas-75-exercises-with-solutions/notebook) with 75 pandas exercises and solutions. You can go along the exercises and compare your solution with the solution in the article.
    
    
    
Take care and good luck! 😊      
</div>

## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ✔️</h2>
    
There's an introduction, which is good. It is important to write an introductory part because it gives an idea about the content of the project. Don't forget about it in the next projects 😊
    
    
</div>

## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
# importing pandas
import pandas as pd

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ⚠️</b>
    
It's better to use **pd** as a reference to pandas. It is convenient as every programmer or analyst has to look for the information on the web or in a documentation. Sometimes you may search for a ready solution and use of a common style is a good idea here as it simplifies understanding someone's code and even copy-paste process. 
     
</div>

<div class='alert alert-info'><b><font color='blue'>I'll swap that out, then. Also fixed it in the cell below, of course. :)</font></b></div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 ✔️</h2>
    
Good :)    
    
</div>

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
# reading the files and storing them to df
df = pd.read_csv("/datasets/movies_and_shows.csv")

Print the first 10 table rows:

In [3]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [4]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ❌</b>
    
 
    
Try to use the `info` method. `describe` is very useful but it gives us the statistics, not general info. 
    

</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 ✔️</h2>
    
Correct!     
</div>

<div class='alert alert-info'><b><font color='blue'>I misunderstood what was asked for; fixed.</font></b></div>

The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [5]:
# the list of column names in the df table
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [6]:
# renaming columns
fixed_columns = list(df.columns)
for i in range(len(fixed_columns)):
    fixed_columns[i] = fixed_columns[i].strip().lower().replace(" ", "_").replace("0", "o")

# zip() iterates over multiple iterables and creates tuples; convenient for dict()
new_columns = dict(zip(list(df.columns), fixed_columns))
df = df.rename(columns=new_columns)

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ✔️</h2>
    
Nice code! However, you do not need to iterate over the dataframe.      
</div>

<div class='alert alert-info'><b><font color='blue'>Thank you! Might I ask what the intended method was for this? A method like this seemed cleaner than just doing them individually, right?</font></b></div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 ✔️</h2>
    
Actually you are right. It is not a problem to have a small loop that iterates over the list of several names. You are 100% correct here. I thought about applying a function, but your approach is very good. 
    
</div>

Check the result. Print the names of the columns once more:

In [7]:
# checking result: the list of column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [8]:
# calculating missing values
df.isna().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [9]:
# dropping rows where columns with title, scores and votes have missing values
df = df.drop(df[df['imdb_score'].isna()].index)
df = df.drop(df[df['imdb_votes'].isna()].index)

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [10]:
# counting missing values
df.isna().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ✔️</h2>
     
Good.      
</div>

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [11]:
# counting duplicate rows
df.duplicated().sum()

6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [12]:
# Produce table with duplicates (with original rows included) and review last 5 rows
df_duplicates = df[df.duplicated(keep=False)]
df_duplicates.tail()
# The description in the next cell says "there are two clear duplicates" but I don't see them? I trust pandas though

name                 character   role    title   type  \
85569       Jessica Cediel           Liliana Navarro  ACTOR  Lokillo  MOVIE   
85570  Javier Gardeaz?­bal  Agust??n "Peluca" Ort??z  ACTOR  Lokillo  MOVIE   
85571        Carla Giraldo            Valery Reinoso  ACTOR  Lokillo  MOVIE   
85572  Ana Mar??a S?­nchez                   Lourdes  ACTOR  Lokillo  MOVIE   
85577         Isabel Gaona                    Cacica  ACTOR  Lokillo  MOVIE   

       release_year      genres  imdb_score  imdb_votes  
85569          2021  ['comedy']         3.8        68.0  
85570          2021  ['comedy']         3.8        68.0  
85571          2021  ['comedy']         3.8        68.0  
85572          2021  ['comedy']         3.8        68.0  
85577          2021  ['comedy']         3.8        68.0

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [13]:
# removing duplicate rows
df = df.drop_duplicates().reset_index(drop=True)

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ✔️</h2>
    
Correct!      
</div>

Check for duplicate rows once more to make sure you have removed all of them:

In [14]:
# checking for duplicates
df.duplicated().sum()

0

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [15]:
# viewing unique type names
type_column = list(df['type'].sort_values().unique())

print(list(type_column))

['MOVIE', 'SHOW', 'movies', 'shows', 'the movie', 'tv', 'tv series', 'tv show', 'tv shows']


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ✔️</h2>
    
Correct.     
</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ⚠️</h2>
    
You do not need to create a new variable though.      
</div>

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [16]:
# function for replacing implicit duplicates
# note that I struggled on this; in most languages, df would be out-of-scope and would need passed in as a parameter
# since we're not doing def main() or anything, that's probably why?
def replace_wrong_show(wrong_shows_list, correct_show="SHOW"):
    df['type'] = df['type'].replace(wrong_shows_list, correct_show)


Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [17]:
# removing implicit duplicates
wrong_shows_list = ['shows', 'show', 'tv show', 'tv shows', 'tv series', 'tv']
replace_wrong_show(wrong_shows_list, "SHOW")

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [18]:
# viewing unique genre names
type_column = list(df['type'].sort_values().unique())
print(type_column)

['MOVIE', 'SHOW', 'movies', 'the movie']


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ✔️</h2>
    
Great. You wrote and  called a function that worked properly.

In general, it can be performed without a function. To do this, we need a list and the same **replace** method. It will look like this:
    
    
    wrong_names = ['shows', 'tv show', 'tv shows', 'tv series', 'tv']
    correct_name = 'SHOW'  

    df['type'] = df['type'].replace(wrong_genres, correct_genre)
    
    
Or, which is the same:
    
    
    df['type'] = df['type'].replace(['shows', 'tv show', 'tv shows', 'tv series', 'tv'], 'SHOW')
</div>

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [19]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name
data_1999 = df[df['release_year'] >= 1999]
data_1999

name            character      role     title       type  \
1661          Jeff Probst       Himself - Host     ACTOR  Survivor       SHOW   
1952     Benicio del Toro  Franky Four Fingers     ACTOR    Snatch      MOVIE   
1953        Dennis Farina           Cousin Avi     ACTOR    Snatch      MOVIE   
1954         Vinnie Jones    Bullet Tooth Tony     ACTOR    Snatch      MOVIE   
1955            Brad Pitt        Mickey O'Neil     ACTOR    Snatch      MOVIE   
...                   ...                  ...       ...       ...        ...   
73854       A??da Morales              Maritza     ACTOR   Lokillo  the movie   
73855    Adelaida Buscato           Mar??a Paz     ACTOR   Lokillo  the movie   
73856  Luz Stella Luengas         Karen Bayona     ACTOR   Lokillo  the movie   
73857        In??s Prieto                Fanny     ACTOR   Lokillo  the movie   
73858      Julian Gaviria              unknown  DIRECTOR   Lokillo  the movie   

       release_year               genres  imdb_score  imdb_votes  
1661           2000          ['reality']         7.4     24687.0  
1952           2000  ['crime', 'comedy']         8.3    841435.0  
1953           2000  ['crime', 'comedy']         8.3    841435.0  
1954           2000  ['crime', 'comedy']         8.3    841435.0  
1955           2000  ['crime', 'comedy']         8.3    841435.0  
...             ...                  ...         ...         ...  
73854          2021           ['comedy']         3.8        68.0  
73855          2021           ['comedy']         3.8        68.0  
73856          2021           ['comedy']         3.8        68.0  
73857          2021           ['comedy']         3.8        68.0  
73858          2021           ['comedy']         3.8        68.0  

[69881 rows x 9 columns]

In [20]:
# repeat conditional indexing so df has only shows (movies are removed as result)
data_1999 = data_1999[data_1999['type'] == "SHOW"]
data_1999.head()

name                character   role      title  type  \
1661       Jeff Probst           Himself - Host  ACTOR   Survivor  SHOW   
2073     Mayumi Tanaka  Monkey D. Luffy (voice)  ACTOR  One Piece  SHOW   
2074      Kazuya Nakai     Roronoa Zoro (voice)  ACTOR  One Piece  SHOW   
2075     Akemi Okamura             Nami (voice)  ACTOR  One Piece  SHOW   
2076  Kappei Yamaguchi            Usopp (voice)  ACTOR  One Piece  SHOW   

      release_year                                             genres  \
1661          2000                                        ['reality']   
2073          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2074          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2075          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2076          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   

      imdb_score  imdb_votes  
1661         7.4     24687.0  
2073         8.8    117129.0  
2074         8.8    117129.0  
2075         8.8    117129.0  
2076         8.8    117129.0

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ✔️</h2>
    
Correct.
    
</div>

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [21]:
# rounding column with scores
data_1999['imdb_score'] = data_1999['imdb_score'].round()
# Doing it this way creates a warning, but seems to work fine. Hmmm.
#checking the outcome with tail()
data_1999.tail()

/tmp/ipykernel_31/1272915053.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_1999['imdb_score'] = data_1999['imdb_score'].round()


name character      role      title  type  release_year  \
73831     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  SHOW          2021   
73832    Rudklao Amratisha   unknown     ACTOR  Let's Eat  SHOW          2021   
73833      Jaturong Mokjok   unknown     ACTOR  Let's Eat  SHOW          2021   
73834     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  SHOW          2021   
73835  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  SHOW          2021   

                    genres  imdb_score  imdb_votes  
73831  ['drama', 'comedy']         8.0         5.0  
73832  ['drama', 'comedy']         8.0         5.0  
73833  ['drama', 'comedy']         8.0         5.0  
73834  ['drama', 'comedy']         8.0         5.0  
73835  ['drama', 'comedy']         8.0         5.0

It is now time to identify outliers based on the number of votes.

In [22]:
# Use groupby() for scores and count all unique values in each group, print the result
# Original answer:
# print(data_1999.groupby('imdb_score').count())

# New answer:
print(data_1999.groupby('imdb_score').nunique())

            name  character  role  title  type  release_year  genres  \
imdb_score                                                             
2.0           24         18     2      5     1             4       5   
3.0           27         27     2      4     1             3       4   
4.0          174        108     2     30     1            13      19   
5.0          581        419     2     94     1            12      68   
6.0         2365       1896     2    347     1            19     212   
7.0         4342       3813     2    593     1            24     304   
8.0         4194       4028     2    539     1            23     329   
9.0          542        498     2     65     1            20      50   
10.0           8          8     1      1     1             1       1   

            imdb_votes  
imdb_score              
2.0                  5  
3.0                  4  
4.0                 30  
5.0                 91  
6.0                329  
7.0                568  
8.0    

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ❌</b>
    
We are asked about the unique values, not the number of rows. 
    

</div>

<div class='alert alert-info'><b><font color='blue'>Is this what you were asking for?</font></b></div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 ✔️</h2>
    
Correct! `nunique` returns us the number of unique values, while `count` returns us the number of rows. Take a look:
    
</div>

In [26]:
# Reviewer's code 2

data_1999.groupby('imdb_score')['name'].agg(['count', 'nunique'])

count  nunique
imdb_score                
4.0           180      174
5.0           592      581
6.0          2494     2365
7.0          4706     4342
8.0          4842     4194
9.0           557      542

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [23]:
# filter dataframe using two conditions (scores to be in the range 4-9)
data_1999 = data_1999[data_1999['imdb_score'] <= 9]
data_1999 = data_1999[data_1999['imdb_score'] >= 4]
# group scores and corresponding average number of votes, reset index and print the result
groups = data_1999.reset_index().groupby(['imdb_score'])
groups = groups.mean()
groups

index  release_year     imdb_votes
imdb_score                                           
4.0         47682.972222   2018.072222    5277.583333
5.0         50642.133446   2018.881757    3143.942568
6.0         48893.382117   2018.627506    3481.717322
7.0         42935.899278   2018.061411    8727.068211
8.0         37703.322594   2017.120611   30299.460967
9.0         31172.134650   2015.287253  126904.109515

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [24]:
# round column with averages
groups = groups.round()
# rename columns
groups = groups.rename(
    columns= {
        'imdb_score': 'Score',
        'imdb_votes': 'Average Votes'
    }
)
# for some reason 'imdb_score' does not get renamed. I'm unclear why
# print dataframe in descending order
print(groups.sort_values(by='imdb_score', ascending=False))

              index  release_year  Average Votes
imdb_score                                      
9.0         31172.0        2015.0       126904.0
8.0         37703.0        2017.0        30299.0
7.0         42936.0        2018.0         8727.0
6.0         48893.0        2019.0         3482.0
5.0         50642.0        2019.0         3144.0
4.0         47683.0        2018.0         5278.0


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment ✔️</h2>
    
> for some reason 'imdb_score' does not get renamed. I'm unclear why
    
    
Because it is not a column. 

</div>

<div class='alert alert-info'><b><font color='blue'>
    Then what is it? And how do I rename it? See <i>print(groups.info())</i> below. <br /> Does running <i>groupby()</i> turn it into an index?<br /><br /><a href=https://stackoverflow.com/questions/19851005/rename-pandas-dataframe-index>This StackOverflow post</a> helped me figure it out. See the next cell.
</font></b></div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 ✔️</h2>
    
Yes, it became index. Let's print it:    
</div>

In [28]:
# Reviewer's code 2

groups

index  release_year  Average Votes
imdb_score                                      
4.0         47683.0        2018.0         5278.0
5.0         50642.0        2019.0         3144.0
6.0         48893.0        2019.0         3482.0
7.0         42936.0        2018.0         8727.0
8.0         37703.0        2017.0        30299.0
9.0         31172.0        2015.0       126904.0

In [29]:
# Reviewer's code 2

groups.index

Float64Index([4.0, 5.0, 6.0, 7.0, 8.0, 9.0], dtype='float64', name='imdb_score')

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 ✔️</h2>
    
You can also reset index:</div>

In [30]:
# Reviewer's code 2

groups.reset_index()

imdb_score    index  release_year  Average Votes
0         4.0  47683.0        2018.0         5278.0
1         5.0  50642.0        2019.0         3144.0
2         6.0  48893.0        2019.0         3482.0
3         7.0  42936.0        2018.0         8727.0
4         8.0  37703.0        2017.0        30299.0
5         9.0  31172.0        2015.0       126904.0

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 ✔️</h2>
    
It will be a column:</div>

In [32]:
# Reviewer's code 2

groups.reset_index().columns

Index(['imdb_score', 'index', 'release_year', 'Average Votes'], dtype='object')

In [25]:
print(groups.info())

print(groups.rename_axis('IMDB Score'))

<class 'pandas.core.frame.DataFrame'>
Float64Index: 6 entries, 4.0 to 9.0
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          6 non-null      float64
 1   release_year   6 non-null      float64
 2   Average Votes  6 non-null      float64
dtypes: float64(3)
memory usage: 192.0 bytes
None
              index  release_year  Average Votes
IMDB Score                                      
4.0         47683.0        2018.0         5278.0
5.0         50642.0        2019.0         3144.0
6.0         48893.0        2019.0         3482.0
7.0         42936.0        2018.0         8727.0
8.0         37703.0        2017.0        30299.0
9.0         31172.0        2015.0       126904.0


<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2 ✔️</h2>
    
It's great that you find solutions and answers to your questions. I would argue it is one of the most important skills. </div>

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.